File: uuProposals.r
# \file      uuProposals.r
# \brief     Functions to handle proposals (only methods that haven't been
#            rewritten in Python yet).
# \copyright Copyright (c) 2019 Utrecht University. All rights reserved.
# \license   GPLv3, see LICENSE.


# \brief Retrieve descriptive information of a number of research proposals.
#        This is used to render a paginated table of research proposals.
#
# \param[in] limit  The number of proposals to return.
# \param[in] offset Offset used for table pagination.
#
# \return List of descriptive information about a number of research proposals.
#
uuGetProposals(*limit, *offset, *result, *status, *statusInfo) {
	*status = "Success";
	*statusInfo = "";

	# Query iRODS to get a list of submitted proposals (i.e. subcollections
	# of the the research-datarequest collection)
	*path = "/tempZone/home/datarequests-research";
	*fields = list("COLL_NAME", "COLL_CREATE_TIME", "COLL_OWNER_NAME", "META_DATA_ATTR_VALUE");
	*conditions = list(uucondition("COLL_PARENT_NAME", "=", *path),
                           uucondition("DATA_NAME", "=", "proposal.json"),
                           uucondition("META_DATA_ATTR_NAME", "=", "status"));
	*orderby = "COLL_NAME";
	*ascdesc = "asc";

	uuPaginatedQuery(*fields, *conditions, *orderby, *ascdesc, *limit, *offset, *kvpList, *status, *statusInfo);
	uuKvpList2JSON(*kvpList, *result, *size);
}


# \brief Same as uuGetProposals, but with a different META_DATA_ATTR_NAME as
#        query condition and fewer return fields. This is a necessary hack,
#        because the iRODS query language does not have an OR operator
#
# \param[in] limit         The number of proposals to return.
# \param[in] offset        Offset used for table pagination.
# \param[in] attributeName The attribute whose value should be returned
#
# \return List of descriptive information about a number of research proposals.
#
uuGetProposalsAdditionalFields(*limit, *offset, *attributeName, *result, *status, *statusInfo) {
	*status = "Success";
	*statusInfo = "";

	*path = "/tempZone/home/datarequests-research";
	*fields = list("COLL_NAME", "META_DATA_ATTR_VALUE");
	*conditions = list(uucondition("COLL_PARENT_NAME", "=", *path),
                           uucondition("DATA_NAME", "=", "proposal.json"),
                           uucondition("META_DATA_ATTR_NAME", "=", *attributeName));
	*orderby = "COLL_NAME";
	*ascdesc = "asc";

	uuPaginatedQuery(*fields, *conditions, *orderby, *ascdesc, *limit, *offset, *kvpList, *status, *statusInfo);
	uuKvpList2JSON(*kvpList, *result, *size);
}

# \brief Request proposal metadata change
#
# \param[in]  proposalColl                  Path to collection of the proposal
# \param[in]  attributeName                 Name of attribute to change
# \param[in]  newAttributeValue             New value of attribute
# \param[in]  newAttributeValueArrayLength  If newAttributeVaue is an array,
#                                           specify length. Needed for parsing
#
requestProposalMetadataChange(*proposalColl, *attributeName,
                              *newAttributeValue, *newAttributeValueArrayLength,
                              *status, *statusInfo) {
        *status = -1;
        *statusInfo = "Internal server error";

        # Get full name of user requesting the metadata change (i.e. the actor)
        *actor = uuClientFullName;

        # Retrieve collection id
        foreach(*row in SELECT COLL_ID WHERE COLL_NAME = *proposalColl
                AND DATA_NAME = 'proposal.json') {
                *collId = *row.COLL_ID;
        }

        # Set the path of the user group to which the proposal belongs.
        # Hardcoded for now as all proposals are stored in the same user group
        *actorGroupPath = '/tempZone/home/datarequests-research';

        # Construct key-value pair (the key specifies the collection on which
        # the metadata change should be applied; the value is a JSON array
        # consisting of the path to the proposal collection, the attribute name
        # the new attribute value and the actor)
        *json_str = "[]";
        *size = 0;
        msi_json_arrayops(*json_str, *proposalColl, "add", *size);
        msi_json_arrayops(*json_str, *attributeName, "add", *size);
        msi_json_arrayops(*json_str, *newAttributeValue, "add", *size);
        msi_json_arrayops(*json_str, *newAttributeValueArrayLength, "add", *size);
        msi_json_arrayops(*json_str, *actor, "add", *size);
        msiString2KeyValPair("", *kvp);
        msiAddKeyVal(*kvp, UUORGMETADATAPREFIX ++ "proposal_action_" ++
                     *collId, *json_str);

        # Set the delayed rule on the actor group (to be picked up and
        # processed when adminProposalActions is called)
        *err = errormsg(msiSetKeyValuePairsToObj(*kvp, *actorGroupPath, "-C"),
                        *msg);
        if (*err < 0) {
                *status = "Unrecoverable";
                *statusInfo = "*err - *msg";
                succeed;
        }

        # Add proposal action status to actor group
        *proposalStatus = UUORGMETADATAPREFIX ++
                             "proposal_status_action_" ++ "*collId=PENDING";
        msiString2KeyValPair(*proposalStatus, *kvp);
        *err = errormsg(msiSetKeyValuePairsToObj(*kvp, *actorGroupPath, "-C"),
                                                 *msg);
        if (*err < 0) {
                *status = "Unrecoverable";
                *statusInfo = "*err - *msg";
                succeed;
        } else {
                *status = "Success";
                *statusInfo = "";
                succeed;
        }
}


# \brief Perform admin operations on the proposal
#
adminProposalActions() {
        msiExecCmd("admin-proposalactions.sh", uuClientFullName, "", "", 0,
                   *out);
}


# \brief Process proposal metadata change request
#
# \param[in] proposalColl                  Collection of the proposal whose
#                                          metadata should be changed
# \param[in] attributeName                 Name of metadata attribute to change
# \param[in] newAttributeValue             The new value of the metadata
#                                          attribute
# \param[in] newAttributeValueArrayLength  The length of the array in case
#                                          newAttributeValue is a JSON array
# \param[in] actor                         The user that has requested the
#                                          change
#
# \return                                  Status and statusInfo (reports the
#                                          success/failure of the processing)
#
uuProposalProcessMetadataChange(*proposalColl, *attributeName,
                                *newAttributeValue,
                                *newAttributeValueArrayLength, *actor,
                                *status, *statusInfo) {
        *status = "Unknown";
        *statusInfo = "An internal error has occurred";

        # Check if user is rodsadmin
        uuGetUserType(uuClientFullName, *userType);
        if (*userType != "rodsadmin") {
                writeLine("stdout", "uuProposalProcessMetadataChange: " ++
                          "Should only be called by a rodsadmin");
                fail;
        }

        # Construct file path of file whose metadata should be changed
        *filePath = *proposalColl ++ "/proposal.json";

        # Grant temporary write ACL
        msiSetACL("default", "admin:write", uuClientFullName, *filePath);

        # Not so elegant way to handle the special case of assigning a proposal
        # for review to one or more DMC members
        if (*attributeName == "assignedForReview") {

                # Check if proposal is already assigned. If so, remove
                # the current assignees
                *alreadyAssigned = false;
                foreach(*row in SELECT META_DATA_ATTR_VALUE WHERE
                                COLL_NAME = *proposalColl AND
                                DATA_NAME = "proposal.json" AND
                                META_DATA_ATTR_NAME = *attributeName) {
                        *alreadyAssigned = true;
                }
                if (*alreadyAssigned) {
                        *err = msi_rmw_avu("-d", *filePath, "assignedForReview",
                                   "%", "%");
                }

                # Convert JSON array of assignees to list
                *assignees = list();
                for (*i = 0;
                     *i < int(*newAttributeValueArrayLength);
                     *i = *i + 1) {
                        *assignee = "";
                        msi_json_arrayops(*newAttributeValue, *assignee,
                                          "get", *i);
                        *assignees = cons(*assignee, *assignees);
                }

                # Set assignedForReview metadata on proposal
                foreach(*assignee in *assignees) {
                        *AttrValStr = *attributeName ++ "=" ++ *assignee;
                        msiString2KeyValPair(*AttrValStr, *Kvp);
                        *err = errormsg(msiAssociateKeyValuePairsToObj(*Kvp,
                                            *filePath, "-d"), *msg);
                        if (*err < 0) {
                                if (*err == -818000) {
                                        *status = "PermissionDenied";
                                        *statusInfo = "User is not " ++
                                                      "permitted to modify " ++
                                                      "this attribute";
                                } else {
                                        *status = "Unrecoverable";
                                        *statusInfo = "*err - *msg";
                                }
                        } else {
                                *status = "Success";
                                *statusInfo = "";
                        }
                }
        } else {
                # Check if requested change isn't already present
                *currentAttributeValue = "";
                foreach(*row in SELECT META_DATA_ATTR_VALUE WHERE
                                COLL_NAME = *proposalColl AND
                                DATA_NAME = "proposal.json" AND
                                META_DATA_ATTR_NAME = *attributeName) {
                        *currentAttributeValue = *row.META_DATA_ATTR_VALUE;
                }
                # If it is, do not apply the requested change
                if (*currentAttributeValue == *newAttributeValue) {
                        *status = "Success";
                        *statusInfo = "";
                # If not, apply the request change
                } else {
                        *AttrValStr = *attributeName ++ "=" ++
                                      *newAttributeValue;
                        msiString2KeyValPair(*AttrValStr, *Kvp);
                        *err = errormsg(msiSetKeyValuePairsToObj(*Kvp,
                                            *filePath, "-d"), *msg);
                        if (*err < 0) {
                                if (*err == -818000) {
                                        *status = "PermissionDenied";
                                        *statusInfo = "User is not " ++
                                                      "permitted to modify " ++
                                                      "this attribute";
                                } else {
                                        *status = "Unrecoverable";
                                        *statusInfo = "*err - *msg";
                                }
                        } else {
                                *status = "Success";
                                *statusInfo = "";
                        }
                }
        }

        # Revoke temporary write ACL
        msiSetACL("default", "admin:null", uuClientFullName, *filePath);
}

File: uuProposals.py
# \file      uuProposal.py
# \brief     Functions to handle research proposals.
# \copyright Copyright (c) 2019 Utrecht University. All rights reserved.
# \license   GPLv3, see LICENSE.

import json
import irods_types
from datetime import datetime
from genquery import (row_iterator, paged_iterator, AS_DICT, AS_LIST)


def uuMetaAdd(callback, objType, objName, attribute, value):
    keyValPair = callback.msiString2KeyValPair(attribute + "=" + value,
                                               irods_types.KeyValPair())['arguments'][1]
    retval = callback.msiSetKeyValuePairsToObj(keyValPair, objName, objType)


def uuMetaAssociate(callback, objType, objName, attribute, value):
    keyValPair = callback.msiString2KeyValPair(attribute + "=" + value,
                                               irods_types.KeyValPair())['arguments'][1]
    retval = callback.msiAssociateKeyValuePairsToObj(keyValPair, objName, objType)


# \brief Assign a research proposal to one or more DMC members for review
#
# \param[in] assignees          JSON-formatted array of DMC members
# \param[in] researchProposalId Unique identifier of the research proposal
#
def assignProposal(callback, assignees, researchProposalId):
    status = -1
    statusInfo = "Internal server error"

    try:
        # Construct research proposal path
        proposalPath = ('/tempZone/home/datarequests-research/' +
                        researchProposalId + '/proposal.json')

        # Remove existing assignedForReview attributes (in case the list of
        # assignees is updated)
        callback.msi_rmw_avu("-d", proposalPath, "assignedForReview", "%", "%")

        # Set assignedForReview metadata on proposal
        for assignee in json.loads(assignees):
            uuMetaAdd(callback, "-d", proposalPath, "assignedForReview",
                      assignee)

        status = 0
        statusInfo = "OK"
    except:
        pass

    return {'status': status, 'statusInfo': statusInfo}


# \brief Persist a research proposal to disk.
#
# \param[in] data    JSON-formatted contents of the research proposal.
#
def submitProposal(callback, data, rei):
    status = -1
    statusInfo = "Internal server error"

    try:
        # Create collection
        zonePath = '/tempZone/home/datarequests-research/'
        timestamp = datetime.now().strftime('%s')
        collPath = zonePath + str(timestamp)
        callback.msiCollCreate(collPath, 1, 0)

        # Write proposal data as JSON to the created collection
        proposalPath = collPath + "/proposal.json"
        ret_val = callback.msiDataObjCreate(proposalPath, "", 0)
        fileDescriptor = ret_val['arguments'][2]
        callback.msiDataObjWrite(fileDescriptor, data, 0)
        callback.msiDataObjClose(fileDescriptor, 0)

        # Set the proposal fields as AVUs on the proposal JSON file
        rule_args = [proposalPath, "-d", "root", data]
        setJsonToObj(rule_args, callback, rei)

        # Set the status metadata field of the proposal to "submitted"
        uuMetaAdd(callback, "-d", proposalPath, "status", "submitted")

        # Set permissions for certain groups on the subcollection
        callback.msiSetACL("recursive", "write",
                           "datarequests-research-datamanagers", collPath)
        callback.msiSetACL("recursive", "write",
                           "datarequests-research-board-of-directors", collPath)

        status = 0
        statusInfo = "OK"
    except:
        pass

    return {'status': status, 'statusInfo': statusInfo}


# \brief Set the status of a submitted research proposal to "approved"
#
# \param[in] researchProposalId Unique identifier of the research proposal.
#
def approveProposal(callback, researchProposalId, currentUserId):
    status = -1
    statusInfo = "Internal server error"

    try:
        # Check if approving user owns the proposal. If so, do not allow
        # approving
        result = isProposalOwner(callback, researchProposalId, currentUserId)
        if result['isProposalOwner']:
            raise Exception()

        # Construct path to the collection of the proposal
        zoneName = ""
        clientZone = callback.uuClientZone(zoneName)['arguments'][0]
        proposalPath = ("/" + clientZone + "/home/datarequests-research/" +
                        researchProposalId + "/proposal.json")
        uuMetaAdd(callback, "-d", proposalPath, "status", "approved")
        status = 0
        statusInfo = "OK"
    except:
        pass

    return {'status': status, 'statusInfo': statusInfo}


# \brief Retrieve a research proposal.
#
# \param[in] researchProposalId Unique identifier of the research proposal.
#
# \return The JSON-formatted contents of the research proposal and the status
#         of the research proposal.
#
def getProposal(callback, researchProposalId):
    status = -1
    statusInfo = "Internal server error"

    # Set collection path and file path
    collPath = '/tempZone/home/datarequests-research/' + researchProposalId
    filePath = collPath + '/proposal.json'

    # Get the size of the proposal JSON file and the status of the proposal
    results = []
    rows = row_iterator(["DATA_SIZE", "META_DATA_ATTR_VALUE"],
                        "COLL_NAME = '%s' and DATA_NAME = '%s' and META_DATA_ATTR_NAME = 'status'"
                        % (collPath, 'proposal.json'),
                        AS_DICT, callback)
    for row in rows:
        dataSize = row['DATA_SIZE']
        proposalStatus = row['META_DATA_ATTR_VALUE']

    # Get the contents of the proposal JSON file
    try:
        ret_val = callback.msiDataObjOpen("objPath=%s" % filePath, 0)
        fileDescriptor = ret_val['arguments'][1]
        ret_val = callback.msiDataObjRead(fileDescriptor, dataSize,
                                          irods_types.BytesBuf())
        fileBuffer = ret_val['arguments'][2]
        callback.msiDataObjClose(fileDescriptor, 0)
        proposalJSON = ''.join(fileBuffer.buf)
        status = 0
        statusInfo = "OK"
    except:
        proposalStatus = ""
        proposalJSON = ""

    return {'proposalJSON': proposalJSON, 'proposalStatus': proposalStatus,
            'status': status, 'statusInfo': statusInfo}


# \brief Get the owner of a research proposal.
#
# \param[in] researchProposalId Unique identifier of the research proposal.
#
# \return The user ID of the owner of the research proposal.
#
def isProposalOwner(callback, researchProposalId, currentUserId):
    status = -1
    statusInfo = "Internal server error"
    isProposalOwner = True

    # Get user ID of proposal owner
    try:
        # Construct path to the collection of the proposal
        zoneName = ""
        clientZone = callback.uuClientZone(zoneName)['arguments'][0]
        collPath = ("/" + clientZone + "/home/datarequests-research/" +
                    researchProposalId)

        # Get list of user IDs with permissions on the proposal and the type of
        # permission they have
        rows = row_iterator(["DATA_ACCESS_USER_ID", "DATA_ACCESS_NAME"],
                            ("DATA_NAME = 'proposal.json' and COLL_NAME like "
                            + "'%s'" % collPath),
                            AS_DICT, callback)

        # Get the user ID with ownership permissions
        proposalOwnerUserId = []
        for row in rows:
            if row["DATA_ACCESS_NAME"] == "own":
                proposalOwnerUserId.append(row["DATA_ACCESS_USER_ID"])

        # Check if exactly 1 owner was found. If not, wipe proposalOwner list
        # and set error status code
        if len(proposalOwnerUserId) != 1:
            status = -2
            statusInfo = ("Not exactly 1 owner found. " +
                          "Something is probably wrong.")
            raise Exception()

        # We only have 1 owner. Set proposalOwner to this owner
        proposalOwnerUserId = proposalOwnerUserId[0]

        # Compare the proposal owner user ID to the user ID of the current user
        isProposalOwner = proposalOwnerUserId == currentUserId

        # Set status to OK
        status = 0
        statusInfo = "OK"
    except:
        pass

    # Return data
    return {'isProposalOwner': isProposalOwner, 'status': status,
            'statusInfo': statusInfo}

# \brief Retrieve descriptive information of a number of research proposals.
#        This is used to render a paginated table of research proposals.
#
# \param[in] limit  The number of proposals to return.
# \param[in] offset Offset used for table pagination.
#
# \return List of descriptive information about a number of research proposals.
#
def DRAFTgetProposals(callback, limit, offset):
    # Query iRODS to get a list of submitted proposals (i.e. subcollections
    # of the datarequests-research collection)
    path = '/tempZone/home/datarequests-research'
    fields = ["COLL_NAME", "COLL_CREATE_TIME", "COLL_OWNER_NAME",
              "META_DATA_ATTR_VALUE"]
    conditions = [callback.uucondition("COLL_PARENT_NAME", "=", path),
                  callback.uucondition("DATA_NAME", "=", "proposal.json")]
    orderby = "COLL_NAME"
    ascdesc = "asc"

    callback.uuPaginatedQuery(fields, conditions, orderby, ascdesc,
                              limit, offset, 0, 0, 0)


def uuSubmitProposal(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(submitProposal(callback,
                                                             rule_args[0], rei)))

def uuIsProposalOwner(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(isProposalOwner(callback,
                                                  rule_args[0], rule_args[1])))

def uuApproveProposal(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(approveProposal(callback,
                                                  rule_args[0], rule_args[1])))


def uuAssignProposal(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(assignProposal(callback,
                                                             rule_args[0], rule_args[1])))


def uuGetProposal(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(getProposal(callback,
                                                          rule_args[0])))


def DRAFTuuGetProposals(rule_args, callback, rei):
    callback.writeString("stdout", json.dumps(getProposals(callback, rule_args[0], rule_args[1])))
